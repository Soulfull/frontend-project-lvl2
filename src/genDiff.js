import fs from 'fs';
import path from 'path';
import compare from './compare.js';
import getParse from './parsers/index.js';
import getFormat from './formatters/index.js';

const genDiff = (paths, formatType) => {
  const items = paths.map((filepath) => {
    const contents = fs.readFileSync(path.resolve(filepath), 'utf8');
    const ext = path.extname(filepath);
    const parse = getParse(ext);
    return parse(contents);
  });
  const diff = compare(...items);
  const format = getFormat(formatType);
  return `${format(diff)}\n`;
};

export default genDiff;
