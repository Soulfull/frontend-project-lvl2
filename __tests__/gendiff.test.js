import path from 'path';
import { fileURLToPath } from 'url';
import { test, expect, jest } from '@jest/globals';
import { genDiff, logDiff } from '../src/genDiff.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const getFixturePath = (filename) => path.join(__dirname, '..', '__fixtures__', filename);

test('diff .json files', () => {
  expect(genDiff([getFixturePath('file1.json'), getFixturePath('file2.json')])).toMatchSnapshot();
});

test('diff .yaml and .yml files', () => {
  expect(genDiff([getFixturePath('file1.yaml'), getFixturePath('file2.yml')])).toMatchSnapshot();
});

test('logDiff. console.log should be called', () => {
  console.log = jest.fn();
  logDiff([getFixturePath('file1.json'), getFixturePath('file2.json')]);
  expect(console.log).toBeCalled();
  expect(console.log.mock.calls[0][0]).toMatchSnapshot();
});

test('diff .json nested files', () => {
  const result = `{
    common: {
      + follow: false
        setting1: Value 1
      - setting2: 200
      - setting3: true
      + setting3: null
      + setting4: blah blah
      + setting5: {
            key5: value5
        }
        setting6: {
            doge: {
              - wow: 
              + wow: so much
            }
            key: value
          + ops: vops
        }
    }
    group1: {
      - baz: bas
      + baz: bars
        foo: bar
      - nest: {
            key: value
        }
      + nest: str
    }
  - group2: {
        abc: 12345
        deep: {
            id: 45
        }
    }
  + group3: {
        deep: {
            id: {
                number: 45
            }
        }
        fee: 100500
    }
}`;
  expect(genDiff([getFixturePath('fileNested1.json'), getFixturePath('fileNested2.json')])).toBe(result);
});
