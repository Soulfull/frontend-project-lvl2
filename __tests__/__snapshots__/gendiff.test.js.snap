// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`diff json format 1`] = `
"[{\\"key\\":\\"common\\",\\"value\\":[{\\"key\\":\\"follow\\",\\"value\\":false,\\"status\\":\\"added\\"},{\\"key\\":\\"setting1\\",\\"value\\":\\"Value 1\\",\\"status\\":\\"equal\\"},{\\"key\\":\\"setting2\\",\\"value\\":200,\\"status\\":\\"removed\\"},{\\"key\\":\\"setting3\\",\\"value\\":true,\\"status\\":\\"updated\\",\\"updatedValue\\":{\\"key\\":\\"value\\"}},{\\"key\\":\\"setting4\\",\\"value\\":\\"blah blah\\",\\"status\\":\\"added\\"},{\\"key\\":\\"setting5\\",\\"value\\":{\\"key5\\":\\"value5\\"},\\"status\\":\\"added\\"},{\\"key\\":\\"setting6\\",\\"value\\":[{\\"key\\":\\"doge\\",\\"value\\":[{\\"key\\":\\"wow\\",\\"value\\":\\"too much\\",\\"status\\":\\"updated\\",\\"updatedValue\\":\\"so much\\"}],\\"status\\":null},{\\"key\\":\\"key\\",\\"value\\":\\"value\\",\\"status\\":\\"equal\\"},{\\"key\\":\\"ops\\",\\"value\\":\\"vops\\",\\"status\\":\\"added\\"}],\\"status\\":null}],\\"status\\":null},{\\"key\\":\\"group1\\",\\"value\\":[{\\"key\\":\\"baz\\",\\"value\\":\\"bas\\",\\"status\\":\\"updated\\",\\"updatedValue\\":\\"bars\\"},{\\"key\\":\\"foo\\",\\"value\\":\\"bar\\",\\"status\\":\\"equal\\"},{\\"key\\":\\"nest\\",\\"value\\":{\\"key\\":\\"value\\"},\\"status\\":\\"updated\\",\\"updatedValue\\":\\"str\\"}],\\"status\\":null},{\\"key\\":\\"group2\\",\\"value\\":{\\"abc\\":12345,\\"deep\\":{\\"id\\":45}},\\"status\\":\\"removed\\"},{\\"key\\":\\"group3\\",\\"value\\":{\\"deep\\":{\\"id\\":{\\"number\\":45}},\\"fee\\":100500},\\"status\\":\\"added\\"},{\\"key\\":\\"group4\\",\\"value\\":[{\\"key\\":\\"default\\",\\"value\\":null,\\"status\\":\\"updated\\",\\"updatedValue\\":\\"\\"},{\\"key\\":\\"foo\\",\\"value\\":0,\\"status\\":\\"updated\\",\\"updatedValue\\":null},{\\"key\\":\\"isNested\\",\\"value\\":false,\\"status\\":\\"updated\\",\\"updatedValue\\":\\"none\\"},{\\"key\\":\\"key\\",\\"value\\":false,\\"status\\":\\"added\\"},{\\"key\\":\\"nest\\",\\"value\\":[{\\"key\\":\\"bar\\",\\"value\\":\\"\\",\\"status\\":\\"updated\\",\\"updatedValue\\":0},{\\"key\\":\\"isNested\\",\\"value\\":true,\\"status\\":\\"removed\\"}],\\"status\\":null},{\\"key\\":\\"someKey\\",\\"value\\":true,\\"status\\":\\"added\\"},{\\"key\\":\\"type\\",\\"value\\":\\"bas\\",\\"status\\":\\"updated\\",\\"updatedValue\\":\\"bar\\"}],\\"status\\":null}]
"
`;
